### PR TITLE
Added lfw dataset

### PR DESCRIPTION
**Pull request by SinghKislay**
Ran on a smaller amount of images as there are a lot of possibilty of triplets from 13,000 images.
Dowload and extract worked just fine. 

tfds.core.DatasetInfo(
    name='lfw',
    version=2.0.0,
    description='Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments',
    urls=['http://vis-www.cs.umass.edu/lfw/#resources'],
    features=FeaturesDict({
        'anchor': Image(shape=(250, 250, 3), dtype=tf.uint8),
        'negative': Image(shape=(250, 250, 3), dtype=tf.uint8),
        'positive': Image(shape=(250, 250, 3), dtype=tf.uint8)
    },
    total_num_examples=69345,
    splits={
        'train': <tfds.core.SplitInfo num_examples=69345>
    },
    supervised_keys=('', ''),
    citation='"""
        @TechReport{LFWTech,
          author = {Gary B. Huang and Manu Ramesh and Tamara Berg and
                          Erik Learned-Miller},
          title = {Labeled Faces in the Wild: A Database for Studying
                          Face Recognition in Unconstrained Environments},
          institution = {University of Massachusetts, Amherst},
          year = 2007,
          number = {07-49},
          month = {October}}
    """',
)

### CODE REVIEW

**Comment by SinghKislay** [comment]
I signed it! the user agreement.

**Comment by SinghKislay** [comment]
@us could you please review the code. Thanks.

File: tensorflow_datasets/image/lfw.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            description=("Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments"),
+            features=tfds.features.FeaturesDict({
+                "anchor": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
+                "positive": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
+                "negative": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
+            }),
+            urls=[self.URL],
+            citation=LFW_CITATION,
+        )
+
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: You should run `pylint` to cleanup unnecessary empty lines. You can check this [link](https://www.tensorflow.org/community/contribute/code_style) about code style.

File: tensorflow_datasets/image/lfw.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    def triplet_maker(self,_path):
+        triplet_list=[]
+        lfw=os.listdir(_path)
+        lfw_mod=[]
+        for lst in lfw:
+            lst_path=os.path.join(_path,lst)
+            temp1=os.listdir(lst_path)
+            if(len(temp1)>1):
+                lfw_mod.append(lst)
+        #print(len(lfw_mod))
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: I think you forget to delete this comment line.

File: tensorflow_datasets/image/lfw.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                            for _it,_i in enumerate(lfw_mod):
+                                path_list_negative=os.path.join(_path,_i)
+                                total_images_negative=os.listdir(path_list_negative)
+                                if(len(total_images_negative)>1):
+                                    if(_it==it):
+                                        break
+                                    else:
+                                        for __i in total_images_negative:
+                                                triplet_dict={'anchor':os.path.join(path,path_list[img_no]),'positive':os.path.join(path,path_list[img_no+offset]),'negative':os.path.join(path_list_negative,__i)}
+                                                #print(triplet_dict)
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: one more forgotten line

File: tensorflow_datasets/image/lfw.py (hunk 2 above)
Comment (us) [inline tensorflow_datasets/image/lfw.py]: one more forgotten line

File: tensorflow_datasets/image/lfw.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+from __future__ import print_function
+
+
+import os
+import tensorflow as tf
+import tensorflow_datasets.public_api as tfds
+
+_URL="http://vis-www.cs.umass.edu/lfw/lfw.tgz"
+LFW_IMAGE_SHAPE=(250,250,3)
+LFW_CITATION="""@TechReport{LFWTech,
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: You can edit your line like others for better readability.
Use `LFW_CITATION="""\`

**Comment by us** [comment]
Thank you for contributing. 😊

**Comment by SinghKislay** [comment]
these images are from the dataset itself. So i don't think they are copyrighted. I'll make the changes regarding the code style asap.

**Comment by us** [comment]
I haven't see any license about this dataset. Please check this [link](https://academia.stackexchange.com/a/63157).

**Comment by SinghKislay** [comment]
@us thank you for your feed-backs. I have changed the images to random generated noise(So, no license issue I hope :) ).  Please do tell if any other changes are required.

**Comment by us** [comment]
Thank you :) But I think you forgot to change the code style.

**Comment by SinghKislay** [comment]
I am sorry but I am a little lost here. I looked at other examples such as catsvsdogs.py and mnist.py. I added the indentations and spaces just like they did and also changed the citation as you mentioned. Could you please point out a little more, Thanks(I am a newbie to open source).

**Comment by us** [comment]
Okay firstly you should check the [pylint](https://www.pylint.org/). It's simply checks your code styles,  detects your errors and rates your code. Then you should read [TensorFlow code style guide](https://www.tensorflow.org/community/contribute/code_style).
 In brief detect errors with pylint and fixed them.

**Comment by SinghKislay** [comment]
@us thank-you for your patience. I have got a pylint score of 4.51(initially it was in negative) also I have followed the Tensorflow guideline of 2 spaces instead of 4. I think it should pass this time.

**Comment by us** [comment]
@SinghKislay but 4.51/10 is a bad score. You should increase the score by checking errors. And I think you understood wrongly. Indentation should be 2 not 4. You are still use 4 space.

**Comment by SinghKislay** [comment]
@us I got a little confused. Now I have got a score of 8.6. And the only errors are too many branches(it is in the helper function, nothing can be done about this) and too few public method and np doc-string for methods(No one added these so I left them as well).

File: tensorflow_datasets/image/lfw.py [hunk 5]
```diff
@@ -40,81 +40,83 @@
...
-            builder=self,
-            description=("Labeled Faces in the Wild:\
-             A Database for Studying Face Recognition in Unconstrained Environments"),
-            features=tfds.features.FeaturesDict({
-              "anchor": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
-              "positive": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
-              "negative": tfds.features.Image(shape=LFW_IMAGE_SHAPE),
+  """LFW Class"""
+  URL = "http://vis-www.cs.umass.edu/lfw/#resources"
+  VERSION = tfds.core.Version("2.0.0")
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: Why do you start from version 2.0.0 ?
Learn more about versioning by checking other datasets commits.

**Comment by us** [comment]
Maybe you can use `# pylint: disable=missing-docstring` to disable the warnings.

File: tensorflow_datasets/image/lfw.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      yield {
+          "anchor": triplet["anchor"],
+          "positive": triplet["positive"],
+          "negative": triplet["negative"],
+      }
+
+  #This is a helper function for making all possible triplets for siamese network(eg. FaceNet)
+  def triplet_maker(self, _path):
+    triplet_list = []
+    lfw = os.listdir(_path)
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: You shuld use `tf.io.gfile` for all filesystem access. Check this [link](https://stackoverflow.com/a/42936969/8401425) about why using that

Comment (SinghKislay) [reply tensorflow_datasets/image/lfw.py]: I understand I have removed them with tf.gfile.ListDirectory

File: tensorflow_datasets/image/lfw.py (hunk 5 above)
Comment (SinghKislay) [reply tensorflow_datasets/image/lfw.py]: Changed the version to 0.1.0

File: tensorflow_datasets/image/lfw.py (hunk 6 above)
Comment (SinghKislay) [reply tensorflow_datasets/image/lfw.py]: Also ran and checked, every thing seems to work fine on my system.

**Comment by cyfra** [comment]
@us - thanks a lot for doing the code review. Appreciated.

@SinghKislay - seems that you're returning the data in the triplet form for Siamese Networks. In tfds we should try to return the dataset in as "raw" format as possible - in this case, it would be just images, as some people might want to use it in other research, where they don't require triplets.
And if needed, people can generate triplets themselves based on the raw image feed.

**Comment by SinghKislay** [comment]
Hello @cyfra I have resolved the issue that you presented. Now the user only gets the image with name of the file. I ran and checked the new code on my system it seems to work. I have included a gist for the dataset_info file here: https://gist.github.com/SinghKislay/e42daac722192c59ddb7f9440702ae64
looking forward to hearing back from you.

File: tensorflow_datasets/image/__init__.py [hunk 7]
```diff
@@ -43,4 +43,6 @@
 from tensorflow_datasets.image.rock_paper_scissors import RockPaperScissors
 from tensorflow_datasets.image.shapes3d import Shapes3d
 from tensorflow_datasets.image.svhn import SvhnCropped
+from tensorflow_datasets.image.voc import Voc2007
 from tensorflow_datasets.image.lfw import LFW
+
```
Comment (us) [inline tensorflow_datasets/image/__init__.py]: empty line

**Comment by us** [comment]
when you do the last changes, maybe you should open new pr that include simple commits. Because you have too much commits to single dataset.

File: tensorflow_datasets/testing/test_data/fake_examples/lfw/lfw/ReadME.txt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+These are auto generated images with random noise
+
+These are structured in just the way it was in original dataset
+
+imgae size= 250 X 250
```
Comment (us) [inline tensorflow_datasets/testing/test_data/fake_examples/lfw/lfw/ReadME.txt]: image

**Comment by SinghKislay** [comment]
@us, do u think everything is okay now.

File: tensorflow_datasets/image/lfw.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    year = 2007,
+    number = {07-49},
+    month = {October}
+}
+"""
+
+
+class LFW(tfds.core.GeneratorBasedBuilder):
+  """LFW Class"""
+  URL = "http://vis-www.cs.umass.edu/lfw/#resources"
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: delete `#resources` in the end.

File: tensorflow_datasets/image/lfw.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  VERSION = tfds.core.Version("0.1.0")
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=("Labeled Faces in the Wild:\
+  A Database for Studying Face Recognition in Unconstrained Environments"),
+        features=tfds.features.FeaturesDict({
+            "image_name": tfds.features.Text(),
```
Comment (us) [inline tensorflow_datasets/image/lfw.py]: `label` could be better thane `image_name` to consistency and meaning.

Comment (SinghKislay) [reply tensorflow_datasets/image/lfw.py]: @us there is no label here in this dataset. it's just dir by the name of a person in which there are pictures of the person, and hence the most logical thing would be to give the name of the image and the image itself. But if you have something else in mind please do suggest.

**Comment by us** [comment]
I think yes. After checking my last reviews, you can close this pr and open a clean new pr.

**Comment by SinghKislay** [comment]
Cool

**Pull request closed without merging**